### PR TITLE
Move the `Comment` class to the new backend interface

### LINKED ISSUE #2223: Migrate `Comment` to the new backend interface

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #2223

### CODE REVIEW

File: aiida/orm/comments.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    def is_stored(self):
+        """
+        Is it already stored or not?
+
+        :return: Boolean
+        """
+        return self._backend_entity.is_stored()
+
+    @abc.abstractproperty
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    def uuid(self):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        return self._backend_entity.uuid
```

File: aiida/orm/comments.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    @abc.abstractmethod
+    def get_ctime(self):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        return self._backend_entity.get_ctime()
```

File: aiida/orm/comments.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    def get_mtime(self):
+        pass
+
+    @abc.abstractmethod
+    def set_mtime(self, value):
+        pass
+
+    @abc.abstractmethod
+    def get_user(self):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        return users.User.from_backend_entity(self._backend_entity.get_user())
```

File: aiida/orm/comments.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    @abc.abstractmethod
+    def set_user(self, value):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        self._backend_entity.set_user(value.backend_entity)
```

File: aiida/orm/comments.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    @abc.abstractmethod
+    def get_content(self):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        return self._backend_entity.get_content()
```

File: aiida/orm/comments.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    def set_ctime(self, value):
+        pass
+
+    @abc.abstractmethod
+    def get_mtime(self):
+        pass
+
+    @abc.abstractmethod
+    def set_mtime(self, value):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        self._backend_entity.set_mtime(value)
```

File: aiida/orm/comments.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    def set_user(self, value):
+        pass
+
+    @abc.abstractmethod
+    def get_content(self):
+        pass
+
+    @abc.abstractmethod
+    def set_content(self, value):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        self._backend_entity.set_content(value)
```

File: aiida/orm/comments.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    def get_ctime(self):
+        pass
+
+    @abc.abstractmethod
+    def set_ctime(self, value):
+        pass
+
+    @abc.abstractmethod
+    def get_mtime(self):
+        pass
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
        return self._backend_entity.get_mtime()
```

File: aiida/orm/comments.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    @abc.abstractproperty
+    def uuid(self):
+        pass
+
+    @abc.abstractmethod
+    def get_ctime(self):
+        pass
+
+    @abc.abstractmethod
+    def set_ctime(self, value):
```
Comment (muhrin) [inline aiida/orm/comments.py]: Why can we set the ctime?  Isn't this just set once at construction?

File: aiida/orm/comments.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+    @abc.abstractmethod
+    def set_user(self, value):
+        pass
+
+    @abc.abstractmethod
+    def get_content(self):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    def get_user(self):
+        pass
+
+    @abc.abstractmethod
+    def set_user(self, value):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    def set_mtime(self, value):
+        pass
+
+    @abc.abstractmethod
+    def get_user(self):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    def get_mtime(self):
+        pass
+
+    @abc.abstractmethod
+    def set_mtime(self, value):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    def set_ctime(self, value):
+        pass
+
+    @abc.abstractmethod
+    def get_mtime(self):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    def get_ctime(self):
+        pass
+
+    @abc.abstractmethod
+    def set_ctime(self, value):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+    @abc.abstractproperty
+    def uuid(self):
+        pass
+
+    @abc.abstractmethod
+    def get_ctime(self):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+        :return: Boolean
+        """
+        return self._backend_entity.is_stored()
+
+    @abc.abstractproperty
+    def uuid(self):
+        pass
+
+    @abc.abstractmethod
```
Comment (muhrin) [inline aiida/orm/comments.py]:
```suggestion
```

File: aiida/orm/comments.py [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+        super(Comment, self).__init__(model)
+
+    def __str__(self):
+        return 'Comment<{}> for node<{}> and user<{}>'.format(self.get_content(), self.node.pk, self.user.email)
+
+    @property
+    def user(self):
+        return users.User.from_backend_entity(self._backend_entity.user)
+
+    def is_stored(self):
```
Comment (muhrin) [inline aiida/orm/comments.py]: This method can be deleted, it's present in the superclass

File: aiida/orm/implementation/django/comment.py [hunk 20]
```diff
@@ -7,126 +7,124 @@
...
+        :param user: a User instance
+        :param content: the comment content
+        :return: a Comment object associated to the given node and user
+        """
+        super(DjangoComment, self).__init__(backend)
+        type_check(user, users.DjangoUser)
+        self._dbmodel = utils.ModelWrapper(DbComment(node=node.dbnode, user=user.dbmodel, content=content))
 
     @property
     def pk(self):
```
Comment (muhrin) [inline aiida/orm/implementation/django/comment.py]: Delete (in superclass)

File: aiida/orm/implementation/django/comment.py [hunk 21]
```diff
@@ -7,126 +7,124 @@
...
-        return self.dbnode.pk
+        return self._dbmodel.pk
 
     @property
     def id(self):
```
Comment (muhrin) [inline aiida/orm/implementation/django/comment.py]: Delete (in superclass)

File: aiida/orm/implementation/django/comment.py [hunk 22]
```diff
@@ -7,126 +7,124 @@
...
-        return self.dbnode.pk
+        return self._dbmodel.pk
 
     @property
-    def to_be_stored(self):
-        return self.dbcomment.pk is None
+    def stored(self):
```
Comment (muhrin) [inline aiida/orm/implementation/django/comment.py]: `is_stored` is already in superclass.  Maybe redirect everything to that and delete this.

File: aiida/orm/implementation/django/comment.py [hunk 23]
```diff
@@ -7,126 +7,124 @@
...
     @property
     def uuid(self):
-        return six.text_type(self.dbcomment.uuid)
+        return six.text_type(self._dbmodel.uuid)
 
     def get_ctime(self):
-        return self.dbcomment.ctime
+        return self._dbmodel.ctime
 
     def set_ctime(self, val):
```
Comment (muhrin) [inline aiida/orm/implementation/django/comment.py]: See comment above about ctime

File: aiida/orm/implementation/sqlalchemy/comment.py [hunk 24]
```diff
@@ -7,121 +7,130 @@
...
+        :param user: a User instance
+        :param content: the comment content
+        :return: a Comment object associated to the given node and user
+        """
+        super(SqlaComment, self).__init__(backend)
+        type_check(user, users.SqlaUser)
+        self._dbmodel = utils.ModelWrapper(DbComment(node=node.dbnode, user=user.dbmodel, content=content))
 
     @property
     def pk(self):
```
Comment (muhrin) [inline aiida/orm/implementation/sqlalchemy/comment.py]: Delete (in superclass)

File: aiida/orm/implementation/sqlalchemy/comment.py [hunk 25]
```diff
@@ -7,121 +7,130 @@
...
-        return self.dbcomment.id
+        return self._dbmodel.pk
 
     @property
     def id(self):
```
Comment (muhrin) [inline aiida/orm/implementation/sqlalchemy/comment.py]: Delete (in superclass)

File: aiida/orm/implementation/sqlalchemy/comment.py [hunk 26]
```diff
@@ -7,121 +7,130 @@
...
-        return self.dbcomment.id
+        return self._dbmodel.pk
 
     @property
-    def to_be_stored(self):
-        return self.dbcomment.id is None
+    def stored(self):
```
Comment (muhrin) [inline aiida/orm/implementation/sqlalchemy/comment.py]: -> `is_stored()`

File: aiida/orm/implementation/sqlalchemy/comment.py [hunk 27]
```diff
@@ -7,121 +7,130 @@
...
     @property
     def uuid(self):
-        return six.text_type(self.dbcomment.uuid)
+        return six.text_type(self._dbmodel.uuid)
 
     def get_ctime(self):
-        return self.dbcomment.ctime
+        return self._dbmodel.ctime
 
     def set_ctime(self, val):
```
Comment (muhrin) [inline aiida/orm/implementation/sqlalchemy/comment.py]: consider deleting

**Comment by sphuber** [comment]
The only "tests" that were in place are indirect through the `verdi comment` command. These are passing, but the coverage is shit and so it doesn't say much. I will see if I can add some proper unit tests for the ORM classes themselves

**Comment by muhrin** [comment]
Good point @giovannipizzi , so long as the interface is clear (i.e. the comment interaction is properly defined through the `Node` interface) then I think it's fine (and maybe even desirable) to not expose this as a first class orm entity.

**Comment by sphuber** [comment]
At the current stage having it as a separate full-fledged entity seems a bit overkill, agree, but I can easily imagine that when we do start to use it, it will be very useful to have it queryable. As I understand its original intention, it is for users to comment on certain nodes, i.e. in a discussion on a platform like MaterialsCloud or just when sharing nodes. For this use case, being able to find all nodes on which a particular user has commented, would be useful. Should we leave it like this for now and I just add tests for the ORM classes?

**Comment by muhrin** [comment]
Yes, leave it as is.  Some testing is good but for now I reckon a 'system testing' approach is fine i.e. just rely on higher level functions like the verdi command line tests to put it through its paces.

**Comment by giovannipizzi** [comment]
Ok for me.
Still, I think we should have at the next meeting a discussion on what should be an Entity and what should be accessible only via Node methods or the QueryBuilder (and yes, they should in the future be queryable via the QB, but I find it more intuitive to do 
```
qb.append(Node, filters={'extras': ..., 'comments': ...})
```
rather
```
qb.append(Node, tag='n')
qb.append(Comment, with_node='n', filters=...)
```
even if maybe it makes the QB code more complex. I'm open to suggestions.
Also mentioning @lekah for comments

**Comment by sphuber** [comment]
I added support for querying `Comment` to the `QueryBuilder` which allowed to clean up the methods on `Node` that deal with comments and made them backend independent.

File: aiida/orm/implementation/django/comment.py [hunk 28]
```diff
@@ -7,126 +7,135 @@
...
+    def mtime(self):
+        return self._dbmodel.mtime
 
-    def set_content(self, val):
-        self.dbcomment.content = val
-        if not self.to_be_stored:
-            self.dbcomment.save()
+    def set_mtime(self, value):
+        self._dbmodel.mtime = value
+        if self.is_stored:
```
Comment (muhrin) [inline aiida/orm/implementation/django/comment.py]: This is not needed.  The `ModelWrapper` does this for all attributes anyway.  Sorry I missed it the first time.

File: aiida/orm/implementation/sqlalchemy/comment.py [hunk 29]
```diff
@@ -7,121 +7,142 @@
...
+    def user(self):
+        return self._backend.users.from_dbmodel(self._dbmodel.user)
 
-    def set_content(self, val):
-        self.dbcomment.content = val
-        if not self.to_be_stored:
-            self.dbcomment.save()
+    def set_user(self, value):
+        self._dbmodel.user = value
+        if self.is_stored:
```
Comment (muhrin) [inline aiida/orm/implementation/sqlalchemy/comment.py]: Not needed

File: aiida/orm/implementation/django/comment.py [hunk 30]
```diff
@@ -7,126 +7,135 @@
...
+        super(DjangoComment, self).__init__(backend)
+        utils.type_check(user, users.DjangoUser)  # pylint: disable=no-member
+        self._dbmodel = ModelWrapper(models.DbComment(dbnode=node.dbnode, user=user.dbmodel, content=content))
 
-    def get_mtime(self):
-        return self.dbcomment.mtime
+    def store(self):
+        """Can only store if both the node and user are stored as well."""
+        if self._dbmodel.dbnode.id is None or self._dbmodel.user.id is None:
+            raise exceptions.ModificationNotAllowed
```
Comment (muhrin) [inline aiida/orm/implementation/django/comment.py]:
```suggestion
            raise exceptions.ModificationNotAllowed('The corresponding node and/or user are not stored')
```

File: aiida/orm/implementation/sqlalchemy/comment.py [hunk 31]
```diff
@@ -7,121 +7,142 @@
...
+        utils.type_check(user, users.SqlaUser)  # pylint: disable=no-member
+        self._dbmodel = ModelWrapper(models.DbComment(dbnode=node.dbnode, user=user.dbmodel, content=content))
 
-    def get_ctime(self):
-        return self.dbcomment.ctime
+    def store(self):
+        """Can only store if both the node and user are stored as well."""
+        if self._dbmodel.dbnode.id is None or self._dbmodel.user.id is None:
+            self._dbmodel.dbnode = None
+            raise exceptions.ModificationNotAllowed
```
Comment (muhrin) [inline aiida/orm/implementation/sqlalchemy/comment.py]:
```suggestion
            raise exceptions.ModificationNotAllowed('The corresponding node and/or user are not stored')
```

File: aiida/orm/implementation/django/comment.py (hunk 30 above)
Comment (muhrin) [reply aiida/orm/implementation/django/comment.py]: Balls, you're breaking 'em

**Pull request merged**